### PR TITLE
Add train_sana_sprint_diffusers file

### PR DESCRIPTION
**Pull request by scxue**
Initial implementation of SANA-Sprint training script adapted for Diffusers.
This needs further refinement and optimization. @lawrence-cj @sayakpaul

### CODE REVIEW

**Comment by sayakpaul** [comment]
Will review in a bit.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 1]
```diff
@@ -0,0 +1,1823 @@
+#!/usr/bin/env python
+# Copyright 2025 The HuggingFace Inc. team. All rights reserved.
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:2]: Feel free to add SANA Sprint team here too :)

File: train_scripts/train_sana_sprint_diffusers.py [hunk 2]
```diff
@@ -0,0 +1,1823 @@
...
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.33.0.dev0")
+
+logger = get_logger(__name__)
+
+if is_torch_npu_available():
+    torch.npu.config.allow_internal_format = False
+
+complex_human_instruction = [
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:87]:
```suggestion
COMPLEX_HUMAN_INSTRUCTION = [
```

File: train_scripts/train_sana_sprint_diffusers.py [hunk 3]
```diff
@@ -0,0 +1,1823 @@
...
+                ) >= self.min_size
+                filter_watermark = (x_json.get("pwatermark", 1.0) or 1.0) <= self.max_pwatermark
+                return filter_size and filter_watermark
+            else:
+                return False
+        except Exception:
+            return False
+
+
+class Text2ImageDataset:
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:162]: Do we have an example dataset with which it would work?

File: train_scripts/train_sana_sprint_diffusers.py [hunk 4]
```diff
@@ -0,0 +1,1823 @@
...
+            self._train_dataset,
+            batch_size=None,
+            shuffle=False,
+            num_workers=num_workers,
+            pin_memory=pin_memory,
+            persistent_workers=persistent_workers,
+        )
+        # add meta-data to dataloader instance for convenience
+        self._train_dataloader.num_batches = num_batches
+        self._train_dataloader.num_samples = num_samples
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:239]: Could use `num_train_examples` here no?

File: train_scripts/train_sana_sprint_diffusers.py [hunk 5]
```diff
@@ -0,0 +1,1823 @@
...
+    widget_dict = []
+    if images is not None:
+        for i, image in enumerate(images):
+            image.save(os.path.join(repo_folder, f"image_{i}.png"))
+            widget_dict.append(
+                {"text": validation_prompt if validation_prompt else " ", "output": {"url": f"image_{i}.png"}}
+            )
+
+    model_description = f"""
+# Sana DreamBooth LoRA - {repo_id}
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:268]:
```suggestion
# Sana SPRINT LoRA - {repo_id}
```

File: train_scripts/train_sana_sprint_diffusers.py [hunk 6]
```diff
@@ -0,0 +1,1823 @@
...
+        prompt=instance_prompt,
+        model_description=model_description,
+        widget=widget_dict,
+    )
+    tags = [
+        "text-to-image",
+        "diffusers-training",
+        "diffusers",
+        "lora",
+        "sana",
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:313]:
```suggestion
        "sana-sprint",
```

File: train_scripts/train_sana_sprint_diffusers.py [hunk 7]
```diff
@@ -0,0 +1,1823 @@
...
+    def train_dataset(self):
+        return self._train_dataset
+
+    @property
+    def train_dataloader(self):
+        return self._train_dataloader
+
+
+# TODO here
+def save_model_card(
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:251]: I think we remove all mentions of LoRA here as it is not doing any LoRA training.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 8]
```diff
@@ -0,0 +1,1823 @@
...
+        args.pretrained_model_name_or_path,
+        subfolder="transformer",
+        revision=args.revision,
+        variant=args.variant,
+        guidance_embeds=False,
+    )
+
+    original_state_dict = load_file(
+        f"{args.pretrained_model_name_or_path}/transformer/diffusion_pytorch_model.safetensors"
+    )
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1103]: I would suggest handling this more robustly. What if the user provides the repo id instead of a local path for `args.pretrained_model_name_or_path`? In that case, this would fail.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 9]
```diff
@@ -0,0 +1,1823 @@
...
+
+    if torch.backends.mps.is_available() and weight_dtype == torch.bfloat16:
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        raise ValueError(
+            "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+        )
+
+    # VAE should always be kept in fp32 for SANA (?)
+    vae.to(dtype=torch.float32)
+    transformer.to(accelerator.device, dtype=weight_dtype)
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1148]: This should not be done as `accelerate.prepare()` will handle things for us.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 10]
```diff
@@ -0,0 +1,1823 @@
...
+            raise ValueError("npu flash attention requires torch_npu extensions and is supported only on npu device ")
+
+    # Initialize a text encoding pipeline and keep it to CPU for now.
+    text_encoding_pipeline = SanaPipeline.from_pretrained(
+        args.pretrained_model_name_or_path,
+        vae=None,
+        transformer=None,
+        text_encoder=text_encoder,
+        tokenizer=tokenizer,
+    )
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1171]:
```suggestion
    # Initialize a text encoding pipeline and keep it to CPU for now.
    text_encoding_pipeline = SanaPipeline.from_pretrained(
        args.pretrained_model_name_or_path,
        vae=None,
        transformer=None,
        text_encoder=text_encoder,
        tokenizer=tokenizer,
        torch_dtype=torch.bfloat16
    )
```

File: train_scripts/train_sana_sprint_diffusers.py [hunk 11]
```diff
@@ -0,0 +1,1823 @@
...
+        # due to pytorch#99272, MPS does not yet support bfloat16.
+        raise ValueError(
+            "Mixed precision training with bfloat16 is not supported on MPS. Please use fp16 (recommended) or fp32 instead."
+        )
+
+    # VAE should always be kept in fp32 for SANA (?)
+    vae.to(dtype=torch.float32)
+    transformer.to(accelerator.device, dtype=weight_dtype)
+    pretrained_model.to(accelerator.device, dtype=weight_dtype)
+    disc.to(accelerator.device, dtype=weight_dtype)
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1150]: Since we're training the heads let's not do this as well.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 12]
```diff
@@ -0,0 +1,1823 @@
...
+        disc.train()
+
+        for step, batch in enumerate(train_dataloader):
+            # text encoding
+            prompts = batch["prompts"]
+            text_encoding_pipeline = text_encoding_pipeline.to("cuda")
+            with torch.no_grad():
+                prompt_embeds, prompt_attention_mask, _, _ = text_encoding_pipeline.encode_prompt(
+                    prompts, complex_human_instruction=complex_human_instruction, do_classifier_free_guidance=False
+                )
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1389]: Let's bring the `text_encoding_pipeline` to CPU once the prompt embeddings are computed?

File: train_scripts/train_sana_sprint_diffusers.py [hunk 13]
```diff
@@ -0,0 +1,1823 @@
...
+                pred = accelerator.unwrap_model(transformer)(
+                    hidden_states=scaled_x_t,
+                    timestep=t.flatten(),
+                    encoder_hidden_states=prompt_embeds,
+                    encoder_attention_mask=prompt_attention_mask,
+                    guidance=(scm_cfg_scale * args.guidance_embeds_scale),
+                    jvp=True,
+                    return_logvar=True,
+                )[0]
+                return pred
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1430]: Why is this needed?

File: train_scripts/train_sana_sprint_diffusers.py [hunk 14]
```diff
@@ -0,0 +1,1823 @@
...
+
+            if phase == "G":
+                transformer.train()
+                disc.eval()
+                models_to_accumulate = [transformer]
+                with accelerator.accumulate(models_to_accumulate):
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1436]:
```suggestion
                with accelerator.accumulate(models_to_accumulate), torch.no_grad():
```

File: train_scripts/train_sana_sprint_diffusers.py [hunk 15]
```diff
@@ -0,0 +1,1823 @@
...
+                    with torch.no_grad():
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1437]: We can then remove this context manager.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 16]
```diff
@@ -0,0 +1,1823 @@
...
+                    accelerator=accelerator,
+                    pipeline_args=pipeline_args,
+                    epoch=epoch,
+                )
+                free_memory()
+
+                images = None
+                del pipeline
+
+    # Save the lora layers
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1754]: We are not doing LoRA. So, this can be safely omitted.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 17]
```diff
@@ -0,0 +1,1823 @@
...
+                disc.eval()
+                models_to_accumulate = [transformer]
+                with accelerator.accumulate(models_to_accumulate):
+                    with torch.no_grad():
+                        cfg_x_t = torch.cat([noisy_model_input, noisy_model_input], dim=0)
+                        cfg_t = torch.cat([t, t], dim=0)
+                        cfg_y = torch.cat([negative_prompt_embeds, prompt_embeds], dim=0)
+                        cfg_y_mask = torch.cat([negative_prompt_attention_mask, prompt_attention_mask], dim=0)
+
+                        cfg_pretrain_pred = pretrained_model(
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1443]: As another optimization, we could keep the `pretrained_model` in CPU once this computation is done and load to GPU again when needed.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 18]
```diff
@@ -0,0 +1,1823 @@
...
+                            optimizer_G.zero_grad(set_to_none=True)
+                            optimizer_D.zero_grad(set_to_none=True)
+                            logger.warning("NaN or Inf detected in grad_norm, skipping iteration...")
+                            continue
+
+                        # switch back to phase G and add global step by one.
+                        phase = "G"
+
+                        optimizer_D.step()
+                        optimizer_D.zero_grad(set_to_none=True)
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1684]: I think `set_to_none` is by default True.

File: train_scripts/train_sana_sprint_diffusers.py [hunk 19]
```diff
@@ -0,0 +1,1823 @@
...
+                            continue
+
+                        # switch phase to D
+                        phase = "D"
+
+                        optimizer_G.step()
+                        lr_scheduler.step()
+                        optimizer_G.zero_grad(set_to_none=True)
+
+            elif phase == "D":
```
Comment (sayakpaul) [inline train_scripts/train_sana_sprint_diffusers.py:1581]: So this alternates between two phases in the _same_ training step, right? If so, I would add a comment. 

Also, should we let the users control the step interval in which the discriminator should be updated? Or not really?

**Review by sayakpaul — COMMENTED** [review]
Looking really promising. I left some comments, LMK if they make sense. 

Additionally, if we could wrap the loss computations for the different phases into different functions, I think that will be easier to read. LMK what you think.

**Comment by sayakpaul** [comment]
Please don't hesitate to ping me for running tests, etc.

**Comment by lawrence-cj** [comment]
Adding here:

https://github.com/huggingface/diffusers/pull/11514

**Comment by lawrence-cj** [comment]
Let's duplicate the files in diffusers here. @scxue 
Refer to: https://github.com/huggingface/diffusers/pull/11514
